crates/sui-rest-api/src/transactions/resolve/literal.rs

```diff
@@ -0,0 +1,787 @@
+                format!(
+                    "literal cannot be resolved into type {}",
+                    std::any::type_name::<T>()
+                ),
+            ))
+        }
+    };
+
+    bcs::serialize_into(buf, &n)?;
+
+    Ok(())
+}
+
+fn resolve_as_address(buf: &mut Vec<u8>, value: &UnresolvedValue) -> Result<()> {
+    let address = match value {
+        // parse as ObjectID to handle the case where 0x is present or missing
+        UnresolvedValue::String(s) => s.parse::<ObjectID>().map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                format!("literal cannot be resolved as bool: {e}"),
+            )
+        })?,
+        UnresolvedValue::Null
+        | UnresolvedValue::Bool(_)
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::Array(_) => {
+            return Err(RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                "literal cannot be resolved into type address",
+            ))
+        }
+    };
+
+    bcs::serialize_into(buf, &address)?;
+
+    Ok(())
+}
+
+fn resolve_as_string(buf: &mut Vec<u8>, value: &UnresolvedValue) -> Result<()> {
+    match value {
+        UnresolvedValue::String(s) => {
+            bcs::serialize_into(buf, s)?;
+        }
+        UnresolvedValue::Bool(_)
+        | UnresolvedValue::Null
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::Array(_) => {
+            return Err(RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                "literal cannot be resolved into string",
+            ))
+        }
+    };
+
+    Ok(())
+}
+
+fn resolve_as_option(buf: &mut Vec<u8>, type_: &Type, value: &UnresolvedValue) -> Result<()> {
+    match value {
+        UnresolvedValue::Null => {
+            buf.push(0);
+        }
+        UnresolvedValue::Bool(_)
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::String(_)
+        | UnresolvedValue::Array(_) => {
+            buf.push(1);
+            resolve_literal_to_type(buf, type_, value)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn resolve_as_vector(buf: &mut Vec<u8>, type_: &Type, value: &UnresolvedValue) -> Result<()> {
+    fn write_u32_as_uleb128(buf: &mut Vec<u8>, mut value: u32) {
+        while value >= 0x80 {
+            // Write 7 (lowest) bits of data and set the 8th bit to 1.
+            let byte = (value & 0x7f) as u8;
+            buf.push(byte | 0x80);
+            value >>= 7;
+        }
+        // Write the remaining bits of data and set the highest bit to 0.
+        buf.push(value as u8);
+    }
+
+    match value {
+        UnresolvedValue::Array(array) => {
+            write_u32_as_uleb128(buf, array.len() as u32);
+            for value in array {
+                resolve_literal_to_type(buf, type_, value)?;
+            }
+        }
+        UnresolvedValue::Bool(_)
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::String(_)
```

Comment:
There is maybe an argument for supporting `String` here, if resolving a `vector<u8>` (i.e. it gets the underlying bytes of the string).